### PR TITLE
feat(Google Photos): Restore hidden 'Back up while charging' toggle

### PR DESCRIPTION
**Pull request by xob0t**
This toggle is hidden by default, when enabled, backups run only when the device is charging.

<details>
  <summary>screenshot</summary>
 
image
  
</details>

### CODE REVIEW

**Comment by oSumAtrIX** [comment]
Why is the toggle hidden?

**Comment by xob0t** [comment]
@oSumAtrIX 
It was an option that Google removed at some point. Just because. Google being Google.

https://thenextweb.com/news/google-photos-backup-charging

But if the user had enabled it before, the flag `backup_prefs_had_backup_only_when_charging_enabled` was set to true to keep it available for you. This path does the same.

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    name = "Show hidden 'Back up while charging' toggle",
+    description = "This toggle is hidden by default, when enabled, backups run only when the device is charging",
+    compatiblePackages = [CompatiblePackage("com.google.android.apps.photos")]
+)
+@Suppress("unused")
+object BackupPreferencesFingerprint : BytecodePatch(
+    setOf(BackupPreferencesFingerprint)
+) {
+    override fun execute(context: BytecodeContext) {
+        // patches 'backup_prefs_had_backup_only_when_charging_enabled' to always be true
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt:23]: Follow this comment style:

```suggestion
        // Patches 'backup_prefs_had_backup_only_when_charging_enabled' to always be true.
```

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        BackupPreferencesFingerprint.result?.let {
+            val chargingPrefStringIndex = it.scanResult.stringsScanResult!!.matches.first().index
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt:25]: There's trailing whitespace. Please format the code

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    setOf(BackupPreferencesFingerprint)
+) {
+    override fun execute(context: BytecodeContext) {
+        // patches 'backup_prefs_had_backup_only_when_charging_enabled' to always be true
+        BackupPreferencesFingerprint.result?.let {
+            val chargingPrefStringIndex = it.scanResult.stringsScanResult!!.matches.first().index    
+            it.mutableMethod.apply {
+                val resultRegister = getInstruction<OneRegisterInstruction>(chargingPrefStringIndex+2).registerA
+                // override result as true
+                addInstruction(chargingPrefStringIndex + 3, "const/4 v$resultRegister, 0x1")
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt:29]: Please add a comment for this offset such as if it is before the return instruction

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/fingerprints/BackupPreferencesFingerprint.kt [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package app.revanced.patches.googlephotos.preferences.fingerprints
+
+import app.revanced.patcher.fingerprint.MethodFingerprint
+
+internal object BackupPreferencesFingerprint : MethodFingerprint(
+    returnType = "Lcom/google/android/apps/photos/backup/data/BackupPreferences;",
+    strings = listOf(
+                "backup_prefs_had_backup_only_when_charging_enabled",
+                ),
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/fingerprints/BackupPreferencesFingerprint.kt:9]: Format the code (Use ktlint extension for intellij idea)

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+import app.revanced.patches.googlephotos.preferences.fingerprints.BackupPreferencesFingerprint
+import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
+
+@Patch(
+    name = "Show hidden 'Back up while charging' toggle",
+    description = "This toggle is hidden by default, when enabled, backups run only when the device is charging",
+    compatiblePackages = [CompatiblePackage("com.google.android.apps.photos")]
+)
+@Suppress("unused")
+object BackupPreferencesFingerprint : BytecodePatch(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt:19]: This name is not correct.

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+import app.revanced.patcher.extensions.InstructionExtensions.getInstruction
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.annotation.CompatiblePackage
+import app.revanced.patcher.patch.annotation.Patch
+import app.revanced.patches.googlephotos.preferences.fingerprints.BackupPreferencesFingerprint
+import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
+
+@Patch(
+    name = "Show hidden 'Back up while charging' toggle",
+    description = "This toggle is hidden by default, when enabled, backups run only when the device is charging",
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt:15]:
```suggestion
    description = "Restores a hidden toggle to only run backups when the device is charging.",
```

File: src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt [hunk 7]
```diff
@@ -0,0 +1,33 @@
...
+import app.revanced.patcher.extensions.InstructionExtensions.addInstruction
+import app.revanced.patcher.extensions.InstructionExtensions.getInstruction
+import app.revanced.patcher.patch.BytecodePatch
+import app.revanced.patcher.patch.annotation.CompatiblePackage
+import app.revanced.patcher.patch.annotation.Patch
+import app.revanced.patches.googlephotos.preferences.fingerprints.BackupPreferencesFingerprint
+import com.android.tools.smali.dexlib2.iface.instruction.OneRegisterInstruction
+
+@Patch(
+    name = "Show hidden 'Back up while charging' toggle",
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/googlephotos/preferences/ShowBackUpWhileChargingPatch.kt:14]:
```suggestion
    name = "Restore hidden 'Back up while charging' toggle",
```

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Comment by oSumAtrIX** [comment]
Uploading is just transferring files over network. Unless there's some weird compression going in client side this shouldn't even begin to be a problem

**Review by oSumAtrIX — APPROVED** [review]

**Pull request merged**